server/src/main/java/org/elasticsearch/action/admin/indices/forcemerge/TransportForceMergeAction.java

```diff
@@ -63,10 +69,19 @@ protected ForceMergeRequest readRequestFrom(StreamInput in) throws IOException {
     }
 
     @Override
-    protected EmptyResult shardOperation(ForceMergeRequest request, ShardRouting shardRouting, Task task) throws IOException {
-        IndexShard indexShard = indicesService.indexServiceSafe(shardRouting.shardId().getIndex()).getShard(shardRouting.shardId().id());
-        indexShard.forceMerge(request);
-        return EmptyResult.INSTANCE;
+    protected void shardOperation(ForceMergeRequest request, ShardRouting shardRouting, Task task,
+                                  ActionListener<TransportBroadcastByNodeAction.EmptyResult> listener) {
+        threadPool.executor(ThreadPool.Names.FORCE_MERGE).execute(ActionRunnable.run(listener,
+            () -> {
+                if (task instanceof CancellableTask && ((CancellableTask)task).isCancelled()) {
```

Comment:
Not right now. Given that there was already a general integration of cancellation in the base class, and that with my change here it might introduce a pitfall for cancelling force-merges in case we ever activated cancellation at the level of `BroadcastRequest` (which is my plan), I decided to keep that here.

Reply:
I think I'd prefer `assert (task instanceof CancellableTask) == false` for now, otherwise we'll forget to remove this `if` when a force-merge shard task becomes cancellable and then we'll wonder why it's here some years later.